### PR TITLE
feat(query): Label Join function support

### PR DESCRIPTION
**Pull request by TanviBhavsar**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Support for prometheus label_join function

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 1]
```diff
@@ -78,3 +79,56 @@ case class LabelReplaceFunction(funcParams: Seq[Any])
...
+    var srcLabelValues = ArrayBuffer[String]()
+
+    srcLabel.
+      foreach(x => if (rangeVectorKey.labelValues.contains(ZeroCopyUTF8String(x)))
+        srcLabelValues += rangeVectorKey.labelValues.get(ZeroCopyUTF8String(x)).get.toString
+      else
+        srcLabelValues += ""
+      )
+
+    val labelJoinValue = srcLabelValues.mkString(separator)
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: This statement can go inside else. Change if condition to length ==0

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 2]
```diff
@@ -78,3 +79,56 @@ case class LabelReplaceFunction(funcParams: Seq[Any])
...
+  }
+
+  def labelJoinImpl(rangeVectorKey: RangeVectorKey): RangeVectorKey = {
+    var srcLabelValues = ArrayBuffer[String]()
+
+    srcLabel.
+      foreach(x => if (rangeVectorKey.labelValues.contains(ZeroCopyUTF8String(x)))
+        srcLabelValues += rangeVectorKey.labelValues.get(ZeroCopyUTF8String(x)).get.toString
+      else
+        srcLabelValues += ""
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Yes prometheus treats label which are not present as empty string. This is an example from robust prometheus
http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=label_join(up%2C%22dst%22%2C%22-%22%2C%22instance%22%2C%22job1%22%2C%22job%22)&g0.tab=1

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 3]
```diff
@@ -78,3 +79,56 @@ case class LabelReplaceFunction(funcParams: Seq[Any])
...
   }
 }
+
+case class LabelJoinFunction(funcParams: Seq[Any])
+  extends MiscellaneousFunction {
+
+  val labelIdentifier: String = "[a-zA-Z_][a-zA-Z0-9_:\\-\\.]*"
+
+  require(funcParams.size >= 2,
+    "expected at least 3 argument(s) in call to label_join")
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: FuncParams does not have metric name. Label join needs 3 arguments from user including metric name so I have kept 3 arguments in error message like what prometheus does. For user metric name is also an argument.

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala [hunk 4]
```diff
@@ -78,3 +79,56 @@ case class LabelReplaceFunction(funcParams: Seq[Any])
...
+
+  def labelJoinImpl(rangeVectorKey: RangeVectorKey): RangeVectorKey = {
+    var srcLabelValues = ArrayBuffer[String]()
+
+    srcLabel.
+      foreach(x => if (rangeVectorKey.labelValues.contains(ZeroCopyUTF8String(x)))
+        srcLabelValues += rangeVectorKey.labelValues.get(ZeroCopyUTF8String(x)).get.toString
+      else
+        srcLabelValues += ""
+      )
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Good suggestion! Made the change

**Comment by TanviBhavsar** [comment]
> Looks great. Good list of test cases. Minor suggestions.

Thanks Vish! I have made the changes

File: query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala (hunk 1 above)
Comment (TanviBhavsar) [reply query/src/main/scala/filodb/query/exec/rangefn/MiscellaneousFunction.scala]: Even if srcLabelValues has length 0, the new label value should have separator. So will have to check for labelJoinValue length

**Pull request merged**